azurelinuxagent/common/logcollector.py

```diff
@@ -0,0 +1,360 @@
+            LogCollector._log_to_results_file("Failed to delete file {0} from archive: {1}".format(file, ustr(e)))
+
+    @staticmethod
+    def _remove_deleted_files_from_archive(final_list_of_files):
+        archive_files = LogCollector._get_list_of_files_in_archive()
+
+        for archive_file in archive_files:
+            file_name = LogCollector._convert_archive_name_to_file_name(archive_file)
+
+            if file_name not in final_list_of_files and file_name.lstrip(os.path.sep) not in final_list_of_files:
+                LogCollector._log_to_results_file("Updating archive, removing deleted file {0}".format(archive_file))
+                LogCollector._delete_file_from_archive(archive_file)
+
+    @staticmethod
+    def _add_file_to_archive(file_name, archive_file_name):
+        with tarfile.open(OUTPUT_ARCHIVE_PATH, "a") as archive:
+            archive.add(file_name, arcname=archive_file_name)
+
+    @staticmethod
+    def _get_file_from_archive(archive_file_name):
+        with tarfile.open(OUTPUT_ARCHIVE_PATH, "r") as archive:
+            try:
+                return archive.getmember(archive_file_name)
+            except KeyError:
+                return None
+
+    @staticmethod
+    def _update_files_in_archive(final_list_of_files):
+        for file_name in final_list_of_files:
+            archive_file_name = LogCollector._convert_file_name_to_archive(file_name)
+            archive_file = LogCollector._get_file_from_archive(archive_file_name)
+
+            if archive_file:
+                # If file is present in the archive, update it if needed (if time last modified or size is different)
+                if LogCollector._is_file_updated(file_name, archive_file):
+                    LogCollector._log_to_results_file("Updating archive, updating file {0}".format(archive_file_name))
+                    LogCollector._delete_file_from_archive(archive_file_name)
+                    LogCollector._add_file_to_archive(file_name, archive_file_name)
+                else:
+                    pass  # nothing to be done, file is archive is the same as file on disk
+            else:
+                # File is not present in the archive, add it
+                LogCollector._log_to_results_file("Updating archive, adding new file {0}".format(archive_file_name))
+                LogCollector._add_file_to_archive(file_name, archive_file_name)
+
+    def _parse_manifest_file(self):
+        files_to_collect = set()
+        manifest_entries = self._read_manifest_file()
+
+        for entry in manifest_entries:
+            # The entry can be one of the four flavours:
+            # 1) ll,/etc/udev/rules.d -- list out contents of the folder and store to results file
+            # 2) echo,### Gathering Configuration Files ### -- print message to results file
+            # 3) copy,/var/lib/waagent/provisioned -- add file to list of files to be collected
+            # 4) diskinfo, -- ignore other commands for now
```

Comment:
Since we are using IID's manifest, it may contain other commands needed for IID, but not needed for log collection. I am ignoring those commands until/if we find they are useful for our purposes. I will clarify the comment in the code.